File: Reorder_List.cpp
/**
 * Definition of ListNode
 * class ListNode {
 * public:
 *     int val;
 *     ListNode *next;
 *     ListNode(int val) {
 *         this->val = val;
 *         this->next = NULL;
 *     }
 * }
 */
class Solution {
public:
    /**
     * @param head: The first node of linked list.
     * @return: void
     */
    void reorderList(ListNode *head) {
        // write your code here
        //left-middle and reverse middle-right
        ListNode *dummy = new ListNode(INT_MIN);
        dummy->next = head;
        ListNode *fast = dummy, *slow = dummy;
        while (fast && fast->next){
            slow = slow->next;
            fast = fast->next->next;
        }
        auto temp = slow;
        slow = slow->next;
        temp->next = nullptr;
        
        
        //slow is the head of the mid-right
        //reverse it
        ListNode *right_dummy = new ListNode(INT_MIN);
        auto new_dummy = right_dummy;
        while (slow){
            auto temp = slow;
            slow = slow->next;
            temp->next = right_dummy->next;
            right_dummy->next = temp;
        }
        //combination
        ListNode *right = new_dummy->next;
        head = dummy->next;
        while (right){
            auto temp = right;
            right = right->next;
            temp->next = head->next;
            head->next = temp;
            head = head->next->next;
        }
        
        head = dummy->next;
    }
};

File: Reorder List.py
"""
Definition of ListNode
class ListNode(object):

    def __init__(self, val, next=None):
        self.val = val
        self.next = next
"""
class Solution:
    """
    @param head: The first node of the linked list.
    @return: nothing
    """
    def reorderList(self, head):
        # write your code here
        
        dummy = ListNode(0)
        dummy.next = head
        slow, fast = dummy, dummy
        while fast.next and fast.next.next:
            fast = fast.next.next
            slow = slow.next
        
        tail, slow.next = slow.next, None
        newhead = None
        while tail:
            tmp, tail = tail, tail.next
            tmp.next = newhead
            newhead = tmp
        head = dummy.next
        ans = dummy
        while head:
            dummy.next, head, dummy = head, head.next, head
            dummy.next, newhead, dummy = newhead, newhead.next, newhead
        return ans.next
